### PR TITLE
Add node highlighting feature.

### PR DESCRIPTION
**Pull request by broken-pen**
Hello! :wave: 

I've been very impressed by this plugin. The tree navigation feels very intuitive, the usage is straight-forward and the code is very clear.

One issue that I've found, however, is that at times it feels a bit difficult to know which TS node I'm currently on. I would often use the node selection just to know whether I've currently selected a function name or the entire function call.

This PR adds a function `highlight_node()` that can be used to briefly flash the node that tree-climber is currently on. It also adds an option `highlight` to all the `goto_*()` functions that flashes the moved-to node. The flash duration can be configured with the option `timeout`, the flash color with the option `higroup`. By default, flashing does not occur when executing a macro. This can be turned on via the option `on_macro`.

The code is largely copied from the neovim builtin `vim.highlight.on_yank()`. This is how the feature is currently configured in my init.lua file:
```lua
local climber = require 'tree-climber'
vim.keymap.set({ 'v', 'o' }, 'in', climber.select_node, { silent = true })
vim.keymap.set('n', '<S-Up>', function() climber.goto_parent { highlight = true, skip_comments = true } end, { silent = true })
vim.keymap.set('n', '<S-Down>', function() climber.goto_child { highlight = true, skip_comments = true } end, { silent = true })
vim.keymap.set('n', '<S-Left>', function() climber.goto_prev { highlight = true, skip_comments = true } end, { silent = true })
vim.keymap.set('n', '<S-Right>', function() climber.goto_next { highlight = true, skip_comments = true } end, { silent = true })
vim.keymap.set('n', 'H', function() climber.highlight_node { timeout = 500 } end, { silent = true })
vim.keymap.set('n', '<.', climber.swap_prev, { silent = true })
vim.keymap.set('n', '>.', climber.swap_next, { silent = true })
```

One thing that I'm not sure of is whether tests should be added — and if yes, what these tests could look like.

I hope the feature is suitable to the project's direction and useful to have merged upstream. I completely understand if not, though. In any case, thank you for the great plugin!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,5 +35,9 @@ vim.keymap.set('n', '<c-j>', require('tree-climber').swap_next, keyopts)
 
 Each function optionally accepts a table with a configuration, for example, `goto_next({ skip_comments = true})`.
 
-The only available option so far is:
+The available options so far are:
 * `skip_comments` (boolean) - ignore comment nodes as if they were not there at all (default: false)
+* `highlight` (boolean) - When moving using the `goto_*()` functions, briefly highlight the new node (default: false)
```
Comment (drybalka) [inline README.md:41]: I think it would be better if "When" was lowercase here and below for consistency

File: README.md [hunk 2]
```diff
@@ -35,5 +35,9 @@ vim.keymap.set('n', '<c-j>', require('tree-climber').swap_next, keyopts)
...
+* `timeout` (number) - When highlighting, the time in ms before highlight is cleared (default: 150)
+* `on_macro` (boolean) - When highlighting, highlight when executing a macro (default: false)
```
Comment (drybalka) [inline README.md]: I think this would be better not "when highlighting, ...", but rather "whether to highlight nodes when executing a macro", or something like that

File: lua/tree-climber.lua [hunk 3]
```diff
@@ -281,4 +281,13 @@ M.select_node = function(options)
   vim.api.nvim_win_set_cursor(0, { end_row, end_col })
 end
 
+M.highlight_node = function(options)
+  local path, parser = get_current_node_path(options)
+  if not path or not parser then
+    return
+  end
+  local node = path[#path]
+  require 'tree-climber.highlight'.highlight_node(node, options)
```
Comment (drybalka) [inline lua/tree-climber.lua]: Can you please add the brackets here around the quotes, like in other requires? I know many people write it like that, but it always confuses me =)

File: lua/tree-climber/highlight.lua [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+local node_ns = vim.api.nvim_create_namespace("tree_climber_hlnode")
+local node_timer
+
+function M.highlight_node(node, opts)
+  vim.validate {
+    node = { node, "userdata", true },
+    opts = {
+      opts,
+      function(t)
+        if t == nil then return true else return type(t) == "table" end
```
Comment (drybalka) [inline lua/tree-climber/highlight.lua:17]: I think this can be simplified to `return t == nil or type(t) == "table"`

File: lua/tree-climber/highlight.lua [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+local ts_utils = require("nvim-treesitter.ts_utils")
+
+local node_ns = vim.api.nvim_create_namespace("tree_climber_hlnode")
+local node_timer
+
+function M.highlight_node(node, opts)
+  vim.validate {
+    node = { node, "userdata", true },
+    opts = {
```
Comment (drybalka) [inline lua/tree-climber/highlight.lua:14]: I think all this can be simplified to `opts = { opts, "table", true }`. Or am I wrong?

File: lua/tree-climber/highlight.lua [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      opts,
+      function(t)
+        if t == nil then return true else return type(t) == "table" end
+      end,
+      true
+    },
```
Comment (drybalka) [inline lua/tree-climber/highlight.lua]: I think all this can be simplified to `opts = { opts, "table", true }`. Or am I wrong?

**Comment by drybalka** [comment]
Great job, I like your contribution very much! Never thought I need it, but now will probably use it myself =) I just made a few minor suggestions, maybe you will find the time to address them soon and then I will merge this PR. 

Maybe also add something like `vim.keymap.set('n', '<c-h>', require('tree-climber').highlight_node, keyopts)` to the suggested mappings?

Regarding the tests I think they are probably unnecessary in this case. I mostly added them to catch some corner cases of the treesitter tree. Your code, however, mostly works with one single node, so that should not be an issue. Of course one can come up with a few tests to check whether the highlight indeed appears and then indeed disappears, but for me it feels a bit contrived in this case.

**Comment by broken-pen** [comment]
Thanks for the thorough review (and the compliment)! I think I've addressed all points. The strange call to `vim.validate()` was mostly because I copied `vim.highlight.on_yank()` without understanding the validate API. Apparently you can either pass `{value, type_name_or_list_of_names, is_optional}` or `{value, validator_func, description}` but no mix of the two. I think the current solution is the simplest.

While having another look at the code, I also made a tiny change to `goto_with()`: it now calls the helper function directly without going through the public `M.highlight_node()`; that should avoid fetching the current node twice for no reason.

Let me know if you find anything new or something you missed earlier. Otherwise thanks for your time and effort!

File: lua/tree-climber.lua (hunk 3 above)
Comment (haolian9) [reply lua/tree-climber.lua]: > Can you please add the brackets here around the quotes, like in other requires? I know many people write it like that, but it always confuses me =)

for this case, i think a stylua.toml could be added to this project. oh, do you use stylua or some other tools  for formatting?

Comment (drybalka) [reply lua/tree-climber.lua]: I haven't heard of stylua before so no, I do not use anything currently. Thanks for the suggestion, I will take a look at it!

**Comment by drybalka** [comment]
Thank you for the contribution! 

One remark though, you amended the changes instead of making a new commit and now I cannot see which corrections you did, so I will have to reread the whole PR again. Not a big problem, I am very grateful for the changes =)

**Comment by broken-pen** [comment]
Ah, thanks for pointing this out! I'm not used to submitting PRs and usually use Git for myself, so I tend to be a bit aggressive wiht history rewriting. I'll keep this in mind for the next PR I'll do to a project. Thanks again for your plugin and the review! :D
